### PR TITLE
Overhaul Macrovision and related protections

### PR DESCRIPTION
**Pull request by TheRogueArchivist**

### CODE REVIEW

File: BinaryObjectScanner.Protection/Macrovision.SafeDisc.cs [hunk 1]
```diff
@@ -37,12 +49,32 @@ internal string SafeDiscCheckPortableExecutable(string file, PortableExecutable
             if (sections == null)
                 return null;
 
+            // Found in Redump entry 57986.
+            bool hintNameTableMatch = pex.ImportHintNameTable?.Any(s => s == "LTDLL_Authenticate") ?? false;
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.SafeDisc.cs:53]: Nitpick: Change `s` to `ihnte` for "Import Hint Name Table Entry"

File: BinaryObjectScanner.Protection/Macrovision.SafeDisc.cs [hunk 2]
```diff
@@ -37,12 +49,32 @@ internal string SafeDiscCheckPortableExecutable(string file, PortableExecutable
...
+            if (strs != null)
+            {
+                // Found in Redump entries 14928, 25579, 32751.
+                if (strs.Any(s => s.Contains("LTDLL_Initialise")))
+                    return "SafeDisc Lite";
+                if (strs.Any(s => s.Contains("LTDLL_Authenticate")))
+                    return "SafeDisc Lite";
+                if (strs.Any(s => s.Contains("LTDLL_Unwrap")))
+                    return "SafeDisc Lite";
+            }
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.SafeDisc.cs:73]: Suggestion: Move Lite checks to their own partial class. IIRC you mentioned that LT and normal were different enough?

File: BinaryObjectScanner.Protection/Macrovision.SafeWrap.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    /// Patents that mention the possible use of SafeWrap:
+    /// https://patentscope.wipo.int/search/en/detail.jsf?docId=US42774631&_cid=P20-LGE2MB-82932-1
+    /// https://patentscope.wipo.int/search/en/detail.jsf?docId=US40713491&_cid=P20-LGE2MB-82932-1
+    /// </summary>
+    public partial class Macrovision
+    {
+        /// <inheritdoc cref="BinaryObjectScanner.Interfaces.IPortableExecutableCheck.CheckPortableExecutable(string, PortableExecutable, bool)"/>
+        internal string SafeWrapCheckPortableExecutable(string file, PortableExecutable pex, bool includeDebug)
+        {
+            // TODO: Figure out what SafeWrap is exactly, and add checks.
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.SafeWrap.cs:40]: _X-Files theme intensifies_

File: BinaryObjectScanner.Protection/Macrovision.cs [hunk 4]
```diff
@@ -228,13 +277,162 @@ internal static string Get00000001TMPVersion(string firstMatchedString, IEnumera
...
         {
             // Get the section strings, if they exist
             if (sectionStrings == null)
                 return null;
 
             // If we don't have the "BoG_" string
+            if (!sectionStrings.Any(s => s.Contains("BoG_")))
+                return null;
+
+            // If we don't have the "BoG_ *90.0&!!  Yy>" string
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.cs:409]: Redundant comment

File: BinaryObjectScanner.Protection/Macrovision.cs [hunk 5]
```diff
@@ -228,13 +277,162 @@ internal static string Get00000001TMPVersion(string firstMatchedString, IEnumera
...
+                        return "4.03.086 / Unknown SafeDisc version";
+                    default:
+                        return $"Unknown Version {version} (Report this to us on GitHub)";
+                }
+            }
+
+            return "Unknown Version (Report this to us on GitHub)";
+        }
+
+        private string CheckNewSectionForProtection(string file, bool includeDebug, List<string> sectionStrings, byte[] sectionRaw)
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.cs:399]: What was the reasoning behind splitting this near-identical code out? Are there cases that both the padding and the section contain `BoG` but only one has a version?

File: BinaryObjectScanner.Protection/Macrovision.SafeDisc.cs (hunk 2 above)
Comment (TheRogueArchivist) [reply BinaryObjectScanner.Protection/Macrovision.SafeDisc.cs:73]: There's a very significant overlap with SD1, most of the path checks have overlap.

File: BinaryObjectScanner.Protection/Macrovision.cs (hunk 5 above)
Comment (mnadareski) [reply BinaryObjectScanner.Protection/Macrovision.cs:399]: As per a discussion in Discord, make a new parameter on the original method and gate the functionality behind that instead.

File: BinaryObjectScanner.Protection/Macrovision.cs [hunk 6]
```diff
@@ -48,32 +52,79 @@ public string CheckPortableExecutable(string file, PortableExecutable pex, bool
...
             // Found in hidden resource of "32bit\Tax02\cdac14ba.dll" in IA item "TurboTax Deluxe Tax Year 2002 for Wndows (2.00R)(Intuit)(2002)(352282)".
-            // TODO: Fix File Description not getting properly pulled for this executable.
             // Known versions:
             // 4.16.050 Windows NT 2002/04/24
             if (name?.Equals("Macrovision RTS Service", StringComparison.OrdinalIgnoreCase) == true)
-                return $"Macrovision RTS Service {pex.FileVersion}";
-
-            // Check for specific indications for individual Macrovision protections.
+                resultsList.Add($"Macrovision RTS Service {pex.FileVersion}");
+            string match;
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.cs:70]: Nit: Odd spacing for `match` here. Prefer to either put it down at the CDS check and use another name within the `if` and `else` blocks. An example would be using `sectionMatch` or something similar.

File: BinaryObjectScanner.Protection/Macrovision.cs [hunk 7]
```diff
@@ -48,32 +52,79 @@ public string CheckPortableExecutable(string file, PortableExecutable pex, bool
...
-            List<string> resultsList = new List<string>();
+                int entryPointIndex = pex.FindEntryPointSectionIndex();
+                string entryPointSectionName = pex.SectionNames[entryPointIndex];
 
-            // Check the header padding
-            string match = CheckSectionForProtection(file, includeDebug, pex.HeaderPaddingStrings, pex.HeaderPaddingData);
-            if (!string.IsNullOrWhiteSpace(match))
-            {
-                resultsList.Add(match);
+                switch(entryPointSectionName)
```
Comment (mnadareski) [inline BinaryObjectScanner.Protection/Macrovision.cs:95]:
```suggestion
                switch (entryPointSectionName)
```

**Review by mnadareski — APPROVED** [review]

**Pull request merged**